src/osd/OSD.h

```diff
@@ -2023,6 +2023,23 @@ class OSD : public Dispatcher,
   // -- placement groups --
   RWLock pg_map_lock; // this lock orders *above* individual PG _locks
   ceph::unordered_map<spg_t, PG*> pg_map; // protected by pg_map lock
+  ceph::unordered_map<spg_t, PG*> pg_map_shadow; // read only copy of map
+  ceph::unordered_map<spg_t, PG*> *pg_map_ptr;   // instantaneous pg_map
+
+  class PG_map_subscriber {
```

Comment:
I guess you did it this way to minimize churn, but it seems like PG_map_subscriber should wrap and hide the pg_map_ptr structure exposing a reference upon dereference by overloading operator* -- that would make it more dificult to use the pointer after the subscriber goes out of scope (I like the RAII adaption though!).